### PR TITLE
allow tag keys to be supplied in .osmTag(Collection) filter

### PR DESCRIPTION
**Pull request by tyrasd**
implements #209

# Changes proposed in this pull request:
## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist
- [x] My code follows the [code-style](https://github.com/GIScience/oshdb/blob/master/CONTRIBUTING.md) rules and I have checked on the [static analyses](https://jenkins.ohsome.org/job/oshdb/view/change-requests/) and [benchmark](https://reports.ohsome.org/oshdb-benchmarks/) (if applicable) results
- [x] I have commented my code
- [x] I have written javadoc (required for public methods)
- [x] I have added sufficient unit tests
- [x] I have made corresponding changes to the [documentation](https://github.com/GIScience/oshdb/tree/master/documentation)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/oshdb/blob/master/CHANGELOG.md)
- [x] I have adjusted the [examples](https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/oshdb-examples) and [benchmark](https://reports.ohsome.org/oshdb-benchmarks/) if necessary

### CODE REVIEW

**Comment by SlowMo24** [comment]
great! can I help somehow to get this merged. I would like to use it. Would this be in 0.6.0-SNAPSHOT or 0.5.6?

File: oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java [hunk 1]
```diff
@@ -139,6 +142,18 @@ public void tagList() throws Exception {
         .timestamps(timestamps1)
         .count();
     assertEquals(5, result.intValue());
+    // only tags
```
Comment (tyrasd) [inline oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: this should probably read "tags and keys mixed"

**Comment by tyrasd** [comment]
> can I help somehow to get this merged.

leaving a code review would be a start :wink: 

> Would this be in 0.6.0-SNAPSHOT or 0.5.6?

I guess these's nothing stopping us from releasing this before 0.6 already

File: oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java (hunk 1 above)
Comment (SlowMo24) [reply oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: great. This is what I asked myself. So I can use them mixed and they will be OR joined. so limiting to "landuse"="*" OR "natural"="forest" is now possible.

Comment (tyrasd) [reply oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: exactly

File: oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/MapReducer.java [hunk 2]
```diff
@@ -586,43 +586,51 @@ protected MapReducer(MapReducer<?> obj) {
...
+  public MapReducer<X> osmTag(Collection<? extends OSMTagInterface> tags) {
     MapReducer<X> ret = this.copy();
-    if (tags.size() == 0) {
+    if (tags.isEmpty()) {
       LOG.warn("Empty tag list. No data will match this filter.");
       ret.preFilters.add(ignored -> false);
       ret.filters.add(ignored -> false);
       return ret;
     }
+    Set<Integer> preKeyIds = new HashSet<>();
```
Comment (SlowMo24) [inline oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/MapReducer.java]: we might add a comment here as to what this Set is about. It took me a while to get the difference to the keyIds

File: oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java [hunk 3]
```diff
@@ -139,6 +142,18 @@ public void tagList() throws Exception {
...
+        .osmTag(Arrays.asList(
+            new OSMTag("highway", "residential"),
+            new OSMTag("highway", "unclassified"),
+            new OSMTagKey("building"))
+        )
+        .osmType(OSMType.WAY)
+        .areaOfInterest(bbox)
+        .timestamps(timestamps1)
+        .count();
+    assertEquals(5+42, result.intValue());
```
Comment (SlowMo24) [inline oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: I would propose another test with AND + OR filters mixed like:
```
    // mixed OR + AND filtering
    result = createMapReducerOSMEntitySnapshot()
        .osmTag(Arrays.asList(
            new OSMTag("highway", "residential"),
            new OSMTag("highway", "unclassified"),
            new OSMTagKey("building"))
        )
        .osmTag("name")
        .osmType(OSMType.WAY)
        .areaOfInterest(bbox)
        .timestamps(timestamps1)
        .count();
    assertEquals(?, result.intValue());
```

Comment (tyrasd) [reply oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: this functionality (that filters can be `AND`-ed together) is already covered by the test below (`tagMultiple`). IMHO it's best to keep tests limited to one specific thing each.

**Pull request merged**